File: app/src/androidTest/java/com/wiredcraft/example/ui/MainActivityTest.java
package com.wiredcraft.example.ui;

import static androidx.test.core.app.ApplicationProvider.getApplicationContext;
import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withClassName;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;

import static org.hamcrest.CoreMatchers.anything;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

import androidx.core.widget.NestedScrollView;
import androidx.test.core.app.ActivityScenario;
import androidx.test.espresso.Espresso;
import androidx.test.espresso.IdlingResource;
import androidx.test.espresso.contrib.RecyclerViewActions;
import androidx.test.espresso.matcher.ViewMatchers;
import androidx.test.ext.junit.rules.ActivityScenarioRule;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.filters.LargeTest;

import com.wiredcraft.example.R;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

@RunWith(AndroidJUnit4.class)
@LargeTest
public class MainActivityTest {

    @Rule
    public ActivityScenarioRule<MainActivity> activityScenarioRule =
            new ActivityScenarioRule<>(MainActivity.class);

    private IdlingResource idlingresource;


    @Before
    public void setUp() {
        activityScenarioRule.getScenario().onActivity(new ActivityScenario.ActivityAction<MainActivity>() {
            @Override
            public void perform(MainActivity activity) {
                idlingresource = activity.getCountingIdlingResource();
            }
        });

        Espresso.registerIdlingResources(idlingresource);
    }

    /**
     * load完成点击某一条，然后能跳转到UserDetailActivity
     */
    @Test
    public void jump2Detail() {
        System.out.println("load finish");
        onView(ViewMatchers.withId(R.id.recycle_view))
                .perform(RecyclerViewActions.actionOnItemAtPosition(5, click()));

        onView(withClassName(is(NestedScrollView.class.getCanonicalName()))).check(matches(anything()));
    }

    @After
    public void release() {
        Espresso.unregisterIdlingResources(idlingresource);
    }
}

File: QR-generator/qrgenerator/app/src/androidTest/java/com/inaction/edward/qrgenerator/MainActivityTest.kt
package com.inaction.edward.qrgenerator

import android.support.test.rule.ActivityTestRule
import android.support.test.runner.AndroidJUnit4
import org.junit.Test

import android.support.test.espresso.Espresso.onView
import android.support.test.espresso.action.ViewActions.click
import android.support.test.espresso.assertion.ViewAssertions.matches
import android.support.test.espresso.matcher.ViewMatchers.*
import org.junit.Rule

import org.junit.runner.RunWith

/**
 * Instrumented test, which will execute on an Android device.
 *
 * See [testing documentation](http://d.android.com/tools/testing).
 */
@RunWith(AndroidJUnit4::class)
class MainActivityTest {

    @Rule
    @JvmField
    val mActivityRule = ActivityTestRule(MainActivity::class.java)


    /**
     * Turn animation off on Developer Options before running this test
     */
    @Test
    fun testFabMenu() {

        onView(withId(R.id.scanFab))
                .check(matches(withAlpha(0f)))
        onView(withId(R.id.generateFab))
                .check(matches(withAlpha(0f)))

        onView(withId(R.id.fab_expand_menu_button))
                .perform(click())

        onView(withId(R.id.scanFab))
                .check(matches(withAlpha(1f)))
        onView(withId(R.id.generateFab))
                .check(matches(withAlpha(1f)))
    }


}
